common/src/main/java/io/netty/util/concurrent/DefaultPromise.java

```diff
@@ -301,8 +322,7 @@ public V getNow() {
      */
     @Override
     public boolean cancel(boolean mayInterruptIfRunning) {
-        if (RESULT_UPDATER.get(this) == null &&
-                RESULT_UPDATER.compareAndSet(this, null, new CauseHolder(new CancellationException()))) {
+        if (RESULT_UPDATER.compareAndSet(this, null, CANCELLATION_CAUSE_HOLDER)) {
```

Comment:
taking this above comment into account I think this should be:

```java
if (RESULT_UPDATER.compareAndSet(this, null, new CauseHolder(new LeanCancellationException())) {
```

And then the above change in `cause()` should are removed all together.

Reply:
@normanmaurer this would defeat the primary purpose of the PR which is to not allocate anything at all when the promise is cancelled. I expect it's very common to cancel a promise/future which is no longer needed, and never subsequently call `cause()` or `get()` on it. For example cancelling a task that was scheduled on the event loop - I think that's the use case in #9522.

Note that `isCancelled()` also still works correctly without having to allocate the exception.

Reply:
Well the question is if cancel() is more often called then cause() as otherwise you allocate a new object on each cause() call

Reply:
My guess is that it would be rare for `cause()` to be called more than once on cancelled promises, especially relative to those cases that it's never called. For other kinds of exception, a new `ExcecutionException` is allocated every time `cause()` is called, and j.u.c. futures allocate a new `CancellationException` every time you call `get()` on a cancelled future (e.g. see source of `j.u.c.FutureTask`.

However if you're concerned about that then it's also easily solvable by just replacing `result` with the new exception wrapped in a `new CauseHolder` before returning from `cause()`. If you're also concerned with possible duplicates then it could be CASed.

Reply:
e.g.
```java
    public Throwable cause() {
        Object result = this.result;
        if (result == CANCELLATION_CAUSE_HOLDER) {
            CancellationException ce = new LeanCancellationException();
            RESULT_UPDATER.lazySet(this, ce);
            return ce;
        }
        return (result instanceof CauseHolder) ? ((CauseHolder) result).cause : null;
    }
```

Reply:
@njhill tis sounds like a good idea! Go for it and let me know once you are done :)

Reply:
Why was the get removed? Isn't it cheaper than CAS?

Reply:
@johnou the get was only there before because creating the exception isn't cheap which is not a problem anymore